### DISCUSSION TITLE
[Item] collect annotations by tag(input tags once)

### DISCUSSION THREAD

**Discussion by windingwind** [windingwind/zotero-better-notes#232] [Note Templates]
Author: @windingwind  
Discuss: https://forums.zotero.org/discussion/comment/414433/#Comment_414433  
> Input tags once and search annotations with these tags in the selected items.

- Item 1
  - Tag 1
    - annotation 1
    - annotation 2
  - Tag 2
     - annotation 3
   - ...
- Item 2
  - Tag 1
    - annotation 4
  - Tag 2
    - annotation 5

![image](https://user-images.githubusercontent.com/33902321/184290015-aa60d42e-7db5-47b9-beaa-b51ff0b69b74.png)

```yml
# This template is specifically for importing/sharing, using better 
# notes 'import from clipboard': copy the content and
# goto Zotero menu bar, click Edit->New Template from Clipboard.  
# Do not copy-paste this to better notes template editor directly.
name: "[item] collect annotations by tag"
content: |-
  // @beforeloop-begin
  ${{
    sharedObj.tags = Zotero.getMainWindow().prompt("Please input tags. Split with ',':", "").split(',');
    return "";
  }}$
  // @beforeloop-end
  // @default-begin
  ${{
    if(!sharedObj.tags){
      return "";
    }
  
    async function getAnnotationsBy(item, filter) {
      const annots = item.getAnnotations().filter(filter);
    return await Zotero.BetterNotes.api.convert.annotations2html(annots, { noteItem: targetNoteItem });
    }
    
    const attachments = Zotero.Items.get(topItem.getAttachments()).filter((i) =>
      i.isPDFAttachment() || i.isSnapshotAttachment() || i.isEPUBAttachment()
    );
    let res = `<h1>${topItem.getField("title")}</h1>`;
    for (let attachment of attachments) {
      res += `<h2>${attachment.getFilename()}</h2>`;
      for (let i in sharedObj.tags) {
        const renderedAnnotations = (
          await getAnnotationsBy(
            attachment,
            (_annot) => _annot.getTags().map(_t=>_t.tag).includes(sharedObj.tags[i])
          )
        );
        if (renderedAnnotations) {
          res += `<h3>${sharedObj.tags[i]}</h3>\n${renderedAnnotations}`;
        }
      }
    }
    return res;
  }}$
  // @default-end
```

**Comment by Crinstttina** [comment]
Thank you for your great work! I'm sorry could you tell me why I always get this error after items are selected? I'm using Zotero 7.
![image](https://github.com/windingwind/zotero-better-notes/assets/69063540/f593c1ac-3e22-4fdf-9590-15fab9efed64)

**Reply by windingwind** [reply]
Update the template

**Comment by citadel-lewis** [comment]
Hi, I'm having the same problem as @Crinstttina - I'm also using Zotero 7. @windingwind do you know what the problem might be?

**Reply by windingwind** [reply]
Update the template

**Comment by windingwind** [comment]
The template is updated for the latest beta version (1.1.4-46)

**Reply by Adammartain** [reply]
I have a problem. After I copy  the template(collect annotations by color) and paste it into template editor, some error occours like this # This template is specifically for importing/sharing, using better # notes 'import from clipboard': copy the content and # goto Zotero menu bar, click Edit->New Template from Clipboard. # Do not copy-paste this to better notes template editor directly. name: "[item] collect annotations by color(sort by color) " content: |- // @author windingwind // @link https://github.com/windingwind/zotero-better-notes/discussions/227 $$# This template is specifically for importing/sharing, using better # notes 'import from clipboard': copy the content and # goto Zotero menu bar, click Edit->New Template from Clipboard. # Do not copy-paste this to better notes template editor directly. name: "[item] collect annotations by color(sort by color) " content: |- // @author windingwind // @link https://github.com/windingwind/zotero-better-notes/discussions/227 $$. I realized I made a mistake, but how to fix it?

**Reply by Adammartain** [reply]
![image](https://github.com/windingwind/zotero-better-notes/assets/56383761/ae6736e6-ae45-44d4-9d64-c6eae8243a6b)
